### DISCUSSION TITLE
How to debug a problem with a module that only occurs when using it with remoting?

### DISCUSSION THREAD

**Discussion by bording** [PowerShell/PowerShell#20876] [Q&A] [answered]
I'm running into a problem with my PowerShell module, and I'm not sure how to get more information about what the problem is.

The module in question is available on the [PowerShell Gallery](https://www.powershellgallery.com/packages/Particular.ServiceControl.Management/5.0.0).

It works fine when imported locally, but if the module is imported in a remoting session, using any of the commands ends up with the following error:

```
Error: Could not load file or assembly 'ServiceControlInstaller.Engine, Version=5.0.0.0, Culture=neutral, PublicKeyToken=null'. Uncaught exception during type initialization. (0x80131534)
```

If the problem was happening locally, I'd be able to attach to the `pwsh` process and debug it in VS. I've done that before, and it works fine.

However, I'm currently at a loss on what to do when remoting is involved. Attaching to the local `pwsh` doesn't help because none of the relevant code is running in that process.

What should I do? Is there some PowerShell-specific way of debugging this? Would setting up the VS [remote debugging tools](https://learn.microsoft.com/en-us/visualstudio/debugger/remote-debugging?view=vs-2022&devlangs=csharp&f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(vs.debug.remote.overview)%3Bk(DevLang-csharp)%26rd%3Dtrue) be the way to go here? If so, what process would I be looking for on the remote machine that would be running the cmdlet?

**Comment by rhubarb-geek-nz** [comment]
> 
> If so, what process would I be looking for on the remote machine that would be running the cmdlet?

How are you invoking the call on the remote machine? Are you using something like `Invoke-Command -HostName overthere -ScriptBlock { xxxx }`

Are you talking to an equivalent machine running the same version of PowerShell when you connect?

If you are using Invoke-Command you should be able able to do

```
[System.Diagnostics.Process]::GetCurrentProcess() | Select-Object -ExpandProperty ID
```

on the remote system before you import the module, so you could attach a debugger before you load the module

**Comment by bording** [comment]
> How are you invoking the call on the remote machine?

The report I got from my customer that had the problem was using `Invoke-Command`. When trying to reproduce the problem, I started with `New-PSSession` / `Enter-PSSession` to establish a remote session.

I then installed and imported the module in the remote session. The import works fine, but then I get the error mentioned when using any of the cmdlets of the module.

> Are you talking to an equivalent machine running the same version of PowerShell when you connect?

Yes, 7.4.0 on both.

**Comment by rhubarb-geek-nz** [comment]
Can they load it on that machine if they log in directly on that machine and use PowerShell 7.4.0 ? Lets see if we can remove the fact that it has anything to do with remoting.

**Reply by bording** [reply]
Yes, if I use the module locally on the remote machine, it works fine. It's only remoting that is the problem.

**Comment by rhubarb-geek-nz** [comment]
Is the $Env:PATH much shorter when remoting rather than in an interactive login/console session? 

Two things we need to determine....

1. is it not finding InstallerEngine\ServiceControlInstaller.Engine.dll ?
2. it is finding InstallerEngine\ServiceControlInstaller.Engine.dll but cannot find its dependencies in the current environment

**Comment by rhubarb-geek-nz** [comment]
I notice the native libraries

```
InstallerEngine\runtimes\linux\lib\net6.0\System.DirectoryServices.Protocols.dll
InstallerEngine\runtimes\win\lib\net6.0\System.DirectoryServices.AccountManagement.dll
InstallerEngine\runtimes\win\lib\net6.0\System.Security.Cryptography.ProtectedData.dll
```

In my experience the module loading under PowerShell is quite different about how it finds the dependent libraries in the tree.

For instance does it cope with the subdirectory "InstallerEngine" and also "InstallerEngine\runtimes" ?

My layouts are quite different, the dependent dlls are in the same directory, and the native dlls are in a much flatter tree

See [SQLiteConnection](https://www.powershellgallery.com/packages/rhubarb-geek-nz.SQLiteConnection/1.0.118.0)

```
osx-arm64\SQLite.Interop.dll.dylib
win-arm64\SQLite.Interop.dll.dll
System.Data.SQLite.dll
osx-x64\SQLite.Interop.dll.dylib
win-x64\SQLite.Interop.dll.dll
linux-arm\SQLite.Interop.dll.so
win-arm\SQLite.Interop.dll.dll
win-x86\SQLite.Interop.dll.dll
linux-arm64\SQLite.Interop.dll.so
SQLiteConnection.dll
linux-x64\SQLite.Interop.dll.so
```

**Comment by bording** [comment]
> Is the $Env:PATH much shorter when remoting rather than in an interactive login/console session?

It's basically identical, but it looks like the local path has `C:\Program Files\PowerShell\7` is listed twice, but it's not when viewing it through remoting.

> For instance does it cope with the subdirectory "InstallerEngine" and also "InstallerEngine\runtimes

Yes, this is handled correctly locally because I'm loading the `InstallerEngine` through an [AssemblyLoadContext](https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl.Management.PowerShell/InstallerEngineAssemblyLoadContext.cs).

**Comment by rhubarb-geek-nz** [comment]
> Yes, this is handled correctly locally because I'm loading the `InstallerEngine` through an [AssemblyLoadContext](https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl.Management.PowerShell/InstallerEngineAssemblyLoadContext.cs).

That might help debugging. Can you test a custom release build with additional Console.Error logging added to the loader?

Are you able to replicate in your own remoting scenario/environment, eg you are not dependent on the customer's machine?

**Comment by bording** [comment]
> Are you able to replicate in your own remoting scenario/environment, eg you are not dependent on the customer's machine?

Yes, I am replicating on my own machines at this point, not the customer's environment.

**Comment by rhubarb-geek-nz** [comment]
Are you using WinRM or SSHD ? Not sure if it makes a difference but I imagine the environment would be quite different.

**Comment by bording** [comment]
> Are you using WinRM or SSHD ? Not sure if it makes a difference but I imagine the environment would be quite different.

WinRM

**Comment by bording** [comment]
> That might help debugging. Can you test a custom release build with additional Console.Error logging added to the loader?

Yes, that would be possible if there's no other way, but my primary motivation for asking the question here was to see if there was some way to actually get PowerShell to surface more information, or otherwise learn how to attach a debugger to the thing.

**Reply by rhubarb-geek-nz** [reply, marked as answer]
If you have the process id from

```
[System.Diagnostics.Process]::GetCurrentProcess() | Select-Object -ExpandProperty ID
```

you should be able to attach a debugger, then hopefully the module load statements will appear in the OutputDebugString window

**Reply by rhubarb-geek-nz** [reply]
Meanwhile, _well done_ for getting the `AssemblyLoadContext` working! I think I need to use one to load some of my database connections.

**Reply by bording** [reply]
Wouldn't that just be the process id of the local `pwsh` though? I don't see how that helps.

**Comment by rhubarb-geek-nz** [comment]
> Wouldn't that just be the process id of the local `pwsh` though? I don't see how that helps.

Not if it is in the scriptblock that you execute on the remote machine

```
Invoke-Command -HostName remotehost -ScriptBlock { 
    $HOST 
    [System.Diagnostics.Process]::GetCurrentProcess() | Select-Object -ExpandProperty ID
    Start-Sleep -Seconds 60
    invoke cmdlet that goes wrong
}
```

**Reply by rhubarb-geek-nz** [reply]
Do you get interactive output, or only get the response when it is complete?

**Reply by bording** [reply]
With `New-PSSession` / `Enter-PSSession` I'm getting an interactive terminal session on the remote machine

**Reply by rhubarb-geek-nz** [reply]
To confirm, get your  process id before you enter the remote session, then get it in the remote session and check they are different

**Reply by bording** [reply]
I was able to confirm that the process id I'm getting is a process from the remote machine, and it corresponds to `wsmprovhost.exe` running on the remote machine.

So, it seems like maybe if I could attach to that process on the remote machine, I'd actually be able to debug the module code running there.

That does seem to mean I either need to try and get the VS remote debugging tools working, or install VS directly on the remote machine and debug it directly there.

**Comment by rhubarb-geek-nz** [comment]
Yes, I would try the [Remote Debugging](https://learn.microsoft.com/en-us/visualstudio/debugger/remote-debugging?view=vs-2022) before installing a full Visual Studio environment, the more software tools you have on the remote system the less it is like a production machine.

You still have the Console.Error ( or write to file on the remote machine ) route if you don't want to put additional software on the remote machine.

**Reply by bording** [reply]
The remote machine in this case is just a development VM, so there's no real concern about putting whatever I need on it get things done.

I do think I'll starting with the Remote Debugging tools first and see how hard it is to get those working.

**Comment by iSazonov** [comment]
I wonder why do you use custom [AssemblyLoadContext](https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl.Management.PowerShell/InstallerEngineAssemblyLoadContext.cs) if PowerShell already does all you need?
I believe you could create module file structure as pointed above and pwsh will find all dlls.
Also you could help pwsh with `RequiredAssemblies` in psd1 file.

**Comment by rhubarb-geek-nz** [comment]
One case for a custom AssemblyLoadContext is if there is a conflict between dependencies.

Another thing to consider is try PowerShell 7.3.10

I have found that the dependency loading in 7.4.0 is not the same and have had problems that resolve when go back to 7.3.10

**Comment by bording** [comment]
I was able to remotely debug the WinRM hosting process, and it turns out it's related to the hosting process not initializing things in a way that lets `AssemblyDependencyResolver` work. This is a rather important thing to support in order to load dependencies properly, so I've opened #20909 detailing my findings.

Thanks @rhubarb-geek-nz for pointing me in the direction of figuring out what process was running things on the remote machine.
